// @spotify/web-scripts/cjs/Tasks/CommitTasks.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const debug_1 = __importDefault(require("debug"));
const cross_spawn_1 = __importDefault(require("cross-spawn"));
const git_cz_1 = require("commitizen/dist/cli/git-cz");
const Paths_1 = require("../Paths");
const dbg = debug_1.default('web-scripts:commit');
function precommitTask(task) {
    const cmd = 'npx';
    const args = [
        '--no-install',
        'lint-staged',
        '--config',
        Paths_1.LINT_STAGED_CONFIG,
        ...task.restOptions,
    ];
    dbg('npx args %o', args);
    return cross_spawn_1.default.sync(cmd, args, {
        env: Object.assign({}, process.env, { WEB_SCRIPTS_SHOULD_FIX: task.fix.toString(), WEB_SCRIPTS_RUN_TESTS: task.tests.toString(), WEB_SCRIPTS_ESLINT_CONFIG: task.eslintConfig, WEB_SCRIPTS_JEST_CONFIG: task.jestConfig, WEB_SCRIPTS_PRETTIER_CONFIG: task.prettierConfig }),
        stdio: 'inherit',
    });
}
exports.precommitTask = precommitTask;
function commitTask(task) {
    dbg('running commitizen commit');
    const cliPath = require
        .resolve('commitizen/package.json')
        .replace('package.json', '');
    return git_cz_1.bootstrap({
        cliPath,
        config: {
            path: task.path,
        },
    }, [null, ...task.restOptions]);
}
exports.commitTask = commitTask;
function commitMsgTask(task) {
    const cmd = 'npx';
    const args = [
        '--no-install',
        'commitlint',
        `--config=${task.config}`,
        `--edit=${process.env.HUSKY_GIT_PARAMS}`,
        ...task.restOptions,
    ];
    dbg('npx args %o', args);
    return cross_spawn_1.default.sync(cmd, args, { stdio: 'inherit' });
}
exports.commitMsgTask = commitMsgTask;
function releaseTask(task) {
    const cmd = 'npx';
    const args = ['--no-install', 'semantic-release', ...task.restOptions];
    dbg('npx args %o', args);
    return cross_spawn_1.default.sync(cmd, args, { stdio: 'inherit' });
}
exports.releaseTask = releaseTask;
